datafusion/core/src/execution/context.rs

```diff
@@ -1437,6 +1438,7 @@ impl SessionState {
             Arc::new(ScalarSubqueryToJoin::new()),
             Arc::new(SubqueryFilterToJoin::new()),
             Arc::new(EliminateFilter::new()),
+            Arc::new(ReduceCrossJoin::new()),
```

Comment:
@alamb having dealt with fickle ordering of optimizer rules before, I'm worried that the implementations could become quite brittle if they are reliant on ordering at all. Do you think it could eventually be a good idea to run them in a loop until no more optimization can be accomplished?

Just throwing a crazy idea out there...

Reply:
I agree that a static ordering is likely to produce non optimal plans in some cases. Running in a loop until a "fixed point" is a reasonable however, then there is a danger of optimizer ping-pong (aka one optimizer pass makes a change and then the other undoes it)

Reply:
Yes, I was struggling with that part. It would be nice if we could measure gradient descent somehow, but maybe that's not possible until we add cost functions everywhere.

Reply:
Gradient descent itself is not directly possible - as that needs differentiable functions / cost functions.

But other (black box) optimization algorithms are possible given some cost function. I think the art of writing good cost functions for queries however can be quite an art 🎨

Reply:
> other (black box) optimization algorithms are possible

Yes, this is what I was trying to say, thank you.